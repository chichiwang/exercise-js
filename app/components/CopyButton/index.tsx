import React, { useState } from 'react';
import classNames from 'classnames';

import CopyIcon from 'app/components/CopyIcon';

import styles from './styles.module.scss';

enum CopyStates {
  ready,
  copying,
  copied,
  errored
}

export enum ColorModes {
  dark = 'dark',
  default = 'dark',
  light = 'light',
}

type CopyButtonProps = {
  className?: string;
  colorMode?: ColorModes,
  textToCopy: string;
};

const ButtonContent = {
  [CopyStates.ready]: (<CopyIcon className={styles.copyIcon} />),
  [CopyStates.copying]: (<div className={styles.buttonContent}>⋯</div>),
  [CopyStates.copied]: (<div className={styles.buttonContent}>✓</div>),
  [CopyStates.errored]: (<div className={styles.buttonContent}>!</div>),
};

const messageDisplayTime = 2000;

function CopyButton({
  className,
  colorMode = ColorModes.default,
  textToCopy,
}: CopyButtonProps) {
  const [copyState, setCopyState] = useState(CopyStates.ready);

  const isButtonDisabled = copyState !== CopyStates.ready;

  function resetButton() {
    setCopyState(CopyStates.ready);
  }

  function writeToClipboard() {
    setCopyState(CopyStates.copying);

    const writePromise = navigator.clipboard.writeText(textToCopy);

    writePromise
      .then(function writeComplete() {
        setCopyState(CopyStates.copied);
        setTimeout(resetButton, messageDisplayTime);
      })
      .catch(function writeFailed(err) {
        console.error(err);
        setCopyState(CopyStates.errored);
        setTimeout(resetButton, messageDisplayTime);
      });
  }

  const wrapperClassName = classNames({
    [styles.copyWrapper]: true,
    [styles.lightMode]: colorMode === ColorModes.light,
    [styles.disabled]: copyState === CopyStates.copying,
    [styles.copied]: copyState === CopyStates.copied,
    [styles.error]: copyState === CopyStates.errored,
  });

  const buttonProps = {
    className: styles.copyButton,
    'aria-label': 'Copy text',
    ...(isButtonDisabled ? {} : {
      onClick: writeToClipboard,
    }),
  };

  return (
    <div className={classNames(styles.copyContainer, className)}>
      <div className={wrapperClassName}>
        <div className={styles.message}>
          <div className={styles.successMessage}>
            <span>Copied!</span>
          </div>
          <div className={styles.errorMessage}>
            <span>Failed!</span>
          </div>
        </div>
        <button type="button" {...buttonProps}>
          { ButtonContent[copyState] }
        </button>
      </div>
    </div>
  );
}

export default CopyButton;
