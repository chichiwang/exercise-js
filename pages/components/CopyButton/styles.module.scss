@use 'sass:map';
@use 'app/var.module';

$padding-unit: map.get(var.$content, 'padding');

$button-size: 1.75rem;
$icon-size: 1.125rem;

.copyPosition {
  position: relative;
  width: $button-size;
  height: $button-size;

  .copyContainer {
    position: absolute;
    display: flex;
    top: 0;
    right: 0;
    font-size: 0;

    .message {
      position: relative;
      height: $button-size;
      user-select: none;
      pointer-events: none;

      .successMessage,
      .errorMessage {
        position: absolute;
        display: flex;
        top: 0;
        right: 0.25rem;
        align-items: center;
        padding: calc($padding-unit / 4) calc($padding-unit / 2);
        height: 100%;
        color: map.get(var.$colors, 'white');
        font-size: 0.75rem;
        border-radius: 0.25rem;
        transform: rotateY(-180deg);
        transform-style: preserve-3d;
        backface-visibility: hidden;
        opacity: 0;
        transition: opacity 1500ms ease-out, transform 850ms cubic-bezier(0.35, 1.64, 0.41, 0.8);
      }

      .successMessage {
        background-color: map.get(var.$colors, 'blue');
      }

      .errorMessage {
        background-color: map.get(var.$colors, 'maroon');
      }
    }

    .copyButton {
      display: flex;
      align-items: center;
      margin: 0;
      padding: calc($padding-unit / 4);
      width: $button-size;
      height: $button-size;
      border-radius: 0.25rem;
      border: 1px solid map.get(var.$colors, 'blue');
      background-color: transparent;
      cursor: pointer;

      .copyIcon {
        width: $icon-size;
        height: $icon-size;
        color: map.get(var.$colors, 'blue');
      }

      .buttonContent {
        width: $icon-size;
        height: $icon-size;
        font-size: $icon-size;
        font-weight: bold;
        line-height: $icon-size;
        color: map.get(var.$colors, 'white');
      }

      &:hover,
      &:focus-visible {
        border: 0;
        background-color: map.get(var.$colors, 'blue');

        .copyIcon {
          color: map.get(var.$colors, 'white');
        }
      }
    }

    &.disabled .copyButton {
      border: 0;
      background-color: map.get(var.$colors, 'grey');
      cursor: auto;

      .copyIcon {
        color: map.get(var.$colors, 'white');
      }
    }

    &.copied {
      .message .successMessage {
        transform: rotateY(0deg);
        opacity: 1;
      }

      .copyButton {
        border: 0;
        background-color: map.get(var.$colors, 'blue');
        cursor: auto;

        .copyIcon {
          color: map.get(var.$colors, 'white');
        }
      }
    }

    &.error {
      .message .errorMessage {
        transform: rotateY(0deg);
        opacity: 1;
      }

      .copyButton {
        border: 0;
        background-color: map.get(var.$colors, 'maroon');
        cursor: auto;

        .copyIcon {
          color: map.get(var.$colors, 'white');
        }
      }
    }
  }
}
