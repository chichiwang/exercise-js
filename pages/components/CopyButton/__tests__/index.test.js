import React from 'react';
import renderer from 'react-test-renderer';
import {
  act,
  fireEvent,
  render,
  screen,
  waitFor,
  within,
} from '@testing-library/react';
import '@testing-library/jest-dom';

import CopyButton from '../index';

jest.useFakeTimers();

describe('CopyButton component', () => {
  const WriteTextResolved = Promise.resolve(true);
  const mockClipboard = {
    writeText: jest.fn().mockReturnValue(WriteTextResolved),
  };

  beforeAll(() => {
    global.navigator.clipboard = mockClipboard;
    jest.spyOn(global, 'setTimeout');
  });

  beforeEach(() => {
    global.navigator.clipboard.writeText.mockClear();
    global.setTimeout.mockClear();
  });

  afterAll(() => {
    delete global.navigator.clipBoard;
    global.setTimeout.mockRestore();
  });

  test('matches the snapshot', () => {
    const tree = renderer.create(
      <CopyButton textToCopy="SecretCode" />,
    ).toJSON();

    expect(tree).toMatchSnapshot();
  });

  test('sends text to clipboard API when clicked', async () => {
    const textToCopy = '(288) 555-0153';

    render(<CopyButton textToCopy={textToCopy} />);

    const $button = screen.getByRole('button');
    const { queryByText: queryButtonText } = within($button);

    expect(queryButtonText('⋯')).not.toBeInTheDocument();
    expect(mockClipboard.writeText).toHaveBeenCalledTimes(0);

    fireEvent.click($button);

    expect(mockClipboard.writeText).toHaveBeenCalledTimes(1);
    expect(mockClipboard.writeText).toHaveBeenCalledWith(textToCopy);
    expect(queryButtonText('⋯')).toBeInTheDocument();

    await act(() => WriteTextResolved);

    expect(queryButtonText('✓')).toBeInTheDocument();

    await act(() => jest.runAllTimers());

    expect(global.setTimeout).toHaveBeenCalledTimes(1);
    expect(queryButtonText('⋯')).not.toBeInTheDocument();
    expect(queryButtonText('✓')).not.toBeInTheDocument();
    expect(queryButtonText('!')).not.toBeInTheDocument();
  });

  describe('clipboard API errors', () => {
    beforeAll(() => {
      jest.spyOn(global.console, 'error');
      global.console.error.mockImplementation(() => {});
    });

    beforeEach(() => {
      global.console.error.mockClear();
    });

    afterAll(() => {
      global.console.error.mockRestore();
    });

    test('logs error to console', async () => {
      const textToCopy = 'EncryptionKey';
      const ClipboardError = new Error('Browser took a poop');
      const WriteTextRejected = Promise.reject(ClipboardError);
      mockClipboard.writeText.mockReturnValueOnce(WriteTextRejected);

      render(<CopyButton textToCopy={textToCopy} />);

      const $button = screen.getByRole('button');
      const { queryByText: queryButtonText } = within($button);

      expect(mockClipboard.writeText).toHaveBeenCalledTimes(0);
      expect(global.console.error).toHaveBeenCalledTimes(0);
      expect(queryButtonText('⋯')).not.toBeInTheDocument();

      fireEvent.click($button);

      expect(queryButtonText('⋯')).toBeInTheDocument();

      try {
        await act(() => WriteTextRejected);
      } catch (err) {
        expect(mockClipboard.writeText).toHaveBeenCalledTimes(1);
        expect(global.console.error).toHaveBeenCalledTimes(1);
        expect(global.console.error).toHaveBeenCalledWith(ClipboardError);
        expect(global.setTimeout).toHaveBeenCalledTimes(1);

        await waitFor(() => expect(queryButtonText('!')).toBeInTheDocument());

        await act(() => jest.runAllTimers());

        expect(queryButtonText('⋯')).not.toBeInTheDocument();
        expect(queryButtonText('✓')).not.toBeInTheDocument();
        expect(queryButtonText('!')).not.toBeInTheDocument();
      }
    });
  });
});
