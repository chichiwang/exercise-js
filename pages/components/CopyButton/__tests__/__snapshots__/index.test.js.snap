// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CopyButton component matches the className snapshot 1`] = `
<div
  className="copyPosition center"
>
  <div
    className="copyContainer"
  >
    <div
      className="message"
    >
      <div
        className="successMessage"
      >
        <span>
          Copied!
        </span>
      </div>
      <div
        className="errorMessage"
      >
        <span>
          Failed!
        </span>
      </div>
    </div>
    <button
      aria-label="Copy text"
      className="copyButton"
      onClick={[Function]}
      type="button"
    >
      <svg
        className="copyIcon"
        height="1em"
        viewBox="0 0 24 24"
        width="1em"
        xmlns="http://www.w3.org/2000/svg"
      >
        <g
          fill="none"
          stroke="currentColor"
          strokeLinecap="round"
          strokeLinejoin="round"
          strokeWidth={2}
        >
          <path
            d="M7 9.667A2.667 2.667 0 0 1 9.667 7h8.666A2.667 2.667 0 0 1 21 9.667v8.666A2.667 2.667 0 0 1 18.333 21H9.667A2.667 2.667 0 0 1 7 18.333z"
          />
          <path
            d="M4.012 16.737A2 2 0 0 1 3 15V5c0-1.1.9-2 2-2h10c.75 0 1.158.385 1.5 1"
          />
        </g>
      </svg>
    </button>
  </div>
</div>
`;

exports[`CopyButton component matches the snapshot 1`] = `
<div
  className="copyPosition"
>
  <div
    className="copyContainer"
  >
    <div
      className="message"
    >
      <div
        className="successMessage"
      >
        <span>
          Copied!
        </span>
      </div>
      <div
        className="errorMessage"
      >
        <span>
          Failed!
        </span>
      </div>
    </div>
    <button
      aria-label="Copy text"
      className="copyButton"
      onClick={[Function]}
      type="button"
    >
      <svg
        className="copyIcon"
        height="1em"
        viewBox="0 0 24 24"
        width="1em"
        xmlns="http://www.w3.org/2000/svg"
      >
        <g
          fill="none"
          stroke="currentColor"
          strokeLinecap="round"
          strokeLinejoin="round"
          strokeWidth={2}
        >
          <path
            d="M7 9.667A2.667 2.667 0 0 1 9.667 7h8.666A2.667 2.667 0 0 1 21 9.667v8.666A2.667 2.667 0 0 1 18.333 21H9.667A2.667 2.667 0 0 1 7 18.333z"
          />
          <path
            d="M4.012 16.737A2 2 0 0 1 3 15V5c0-1.1.9-2 2-2h10c.75 0 1.158.385 1.5 1"
          />
        </g>
      </svg>
    </button>
  </div>
</div>
`;
